### PR TITLE
[ENH] html display of estimators - tag and consistent handling of via `_steps_attr` in `HeterogenousMetaEstimator` descendants

### LINKED ISSUE #7152: [BUG] HTML view of transformations are not showing as a flow and instead shown as class definition directly
## Code

Reused from https://github.com/sktime/sktime/discussions/5903 (I still have that question, need to resolve that sometime)

```python
from sklearn.base import BaseEstimator
from sklearn.preprocessing import RobustScaler, QuantileTransformer
from sktime.forecasting.arima import ARIMA
from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
from sktime.forecasting.compose import (
    ForecastingPipeline,
    ForecastX,
    TransformedTargetForecaster,
)
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.base import BaseTransformer
from sktime.transformations.compose import FeatureUnion, TransformerPipeline
from sktime.transformations.series.boxcox import BoxCoxTransformer, LogTransformer
from sktime.transformations.series.detrend import Detrender, Deseasonalizer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.subset import ColumnSelect


def create_pipeline(
    endogenous_forecaster: BaseForecaster,
    prediction_horizon: ForecastingHorizon,
    future_known_exogenous: list[str] | None = None,
    future_unknown_exogenous: list[str] | None = None,
    future_unknown_forecaster: BaseForecaster | None = None,
    endogenous_transformations: list[BaseTransformer | BaseEstimator] | None = None,
    future_known_transformations: list[BaseTransformer | BaseEstimator] | None = None,
    future_unknown_transformations: list[BaseTransformer | BaseEstimator] | None = None,
) -> BaseForecaster:
    if not endogenous_transformations:
        y_pipe = endogenous_forecaster
    else:
        for counter, endogenous_transformation in enumerate(endogenous_transformations):
            if isinstance(endogenous_transformation, BaseTransformer):
                continue

            endogenous_transformations[counter] = TabularToSeriesAdaptor(endogenous_transformation)

        y_pipe = TransformedTargetForecaster(
            [TransformerPipeline(endogenous_transformations), endogenous_forecaster]
        )

    if not future_known_exogenous:
        known_X_pipe = None
    elif not future_known_transformations:
        known_X_pipe = future_known_transformations
    else:
        for counter, future_known_transformation in enumerate(future_known_transformations):
            if isinstance(future_known_transformation, BaseTransformer):
                continue

            future_known_transformations[counter] = TabularToSeriesAdaptor(
                endogenous_transformation
            )

        known_X_pipe = TransformerPipeline(
            [ColumnSelect(columns=future_known_exogenous), *future_known_transformations]
        )

    if not future_unknown_exogenous:
        unknown_X_trafo_pipe = None
    elif not future_unknown_transformations:
        unknown_X_trafo_pipe = None
    else:
        for counter, future_unknown_transformation in enumerate(future_unknown_transformations):
            if isinstance(future_unknown_transformation, BaseTransformer):
                continue

            future_unknown_transformations[counter] = TabularToSeriesAdaptor(
                endogenous_transformation
            )

        unknown_X_trafo_pipe = TransformerPipeline(
            [ColumnSelect(columns=future_unknown_exogenous), *future_unknown_transformations]
        )

    if not future_unknown_exogenous:
        unknown_X_fcst_pipe = None
    elif not future_unknown_forecaster:
        unknown_X_fcst_pipe = endogenous_forecaster
    else:
        unknown_X_fcst_pipe = future_unknown_forecaster

    e2e_trafo_pipe_steps: list[BaseTransformer] = []

    if known_X_pipe and not unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(known_X_pipe)
    elif not known_X_pipe and unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(unknown_X_trafo_pipe)
    elif known_X_pipe and unknown_X_trafo_pipe:
        e2e_trafo_pipe_steps.append(FeatureUnion([known_X_pipe, unknown_X_trafo_pipe]))

    if e2e_trafo_pipe_steps:
        e2e_trafo_pipe = TransformerPipeline(e2e_trafo_pipe_steps)
    else:
        e2e_trafo_pipe = None

    if not unknown_X_fcst_pipe:
        e2e_fcst_pipe = y_pipe
    else:
        e2e_fcst_pipe = ForecastX(
            y_pipe,
            unknown_X_fcst_pipe,
            fh_X=prediction_horizon,
            forecaster_X_exogeneous="complement",
        )

    if e2e_trafo_pipe:
        e2e_pipe = ForecastingPipeline([e2e_trafo_pipe, e2e_fcst_pipe])
    else:
        e2e_pipe = e2e_fcst_pipe

    return e2e_pipe


create_pipeline(
    ARIMA(),
    ForecastingHorizon(values=range(1, 4)),
    future_known_exogenous=["alpha", "beta"],
    future_unknown_exogenous=["gamma", "delta"],
    future_unknown_forecaster=ExponentialSmoothing(),
    endogenous_transformations=[
        LogTransformer(),
        Detrender(model="additive"),
        Deseasonalizer(model="additive"),
    ],
    future_known_transformations=[RobustScaler(), BoxCoxTransformer()],
    future_unknown_transformations=[QuantileTransformer(), Differencer()],
)
```

Result

image

## Expectation

Not sure if it's a bug, but I was expecting a DAG flow for `TransformerPipeline` component.

### PR DESCRIPTION
**Pull request by mateuszkasprowicz**
#### Reference Issues/PRs

Implements #7222 
Fixes #7152 

#### What does this implement/fix? Explain your changes.

- update `sk_visual_block` method of `_HeterogenousMetaestimator` to be based on the `visual_block_kind` tag
- add `visual_block_kind` to tags registry
- add new `_steps` and `steps_` properties to:
     - `ParamFitterPipeline`
     - `PwTrafoPanelPipeline`
     - `RegressorPipeline`
     - `ClassifierPipeline`
     - `ClustererPipeline`
     - `PwTrafoPanelPipeline`
     - `GroupbyCategoryForecaster` 
     - `Permute`
- change `FeatureUnion`'s visual_block_kind tag to parallel and update `_steps_attr` to correct format (an iterable of (name: str, estimator) tuples)
- update `_steps_attr` of `GroupbyCategoryForecaster` and `TransformSelectForecaster` to point to default (the steps properties). Change their `visual_block_kind` tag to parallel.
- update `_steps_attr` of ColumnEnsembleTransformer from "transformers" to "_transformers" which has a correct format (an iterable of (name: str, estimator) tuples) instead of (name: str, estimator, column). Update its `visual_block_kind` tag to parallel.

#### Does your contribution introduce a new dependency? If yes, which one?
No

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by mateuszkasprowicz** [comment]
> 1. Pipelines with a single estimator and transformer(s) (e.g., ParamFitterPipeline, PwTrafoPanelPipeline). These classes inherit _steps_attr, but it points to _steps, which does not exist. They require a new _steps attribute that combines transformers and the estimator.

Implemented part 1

File: sktime/base/_meta.py [hunk 1]
```diff
@@ -713,6 +714,20 @@ def _tagchain_is_linked_set(
...
             self.set_tags(**{mid_tag_name: mid_tag_val_not})
 
+    def _sk_visual_block_(self):
+        steps = getattr(self, self._steps_attr)
+
+        names, estimators = zip(*steps)
+
+        name_details = [str(est) for est in estimators]
+        return _VisualBlock(
+            "serial",
```
Comment (fkiraly) [inline sktime/base/_meta.py:724]: can we make this depend on a tag or attribute?

File: sktime/classification/compose/_pipeline.py [hunk 2]
```diff
@@ -161,6 +161,16 @@ def _transformers(self):
...
         self.transformers_._steps = value
 
+    @property
+    def _steps(self):
+        return self._check_estimators(self.transformers) + [
+            self._coerce_estimator_tuple(self.classifier)
+        ]
+
+    @property
+    def steps_(self):
```
Comment (fkiraly) [inline sktime/classification/compose/_pipeline.py:171]: question, I do not know the answer - do we need `steps_`?

File: sktime/base/_meta.py (hunk 1 above)
Comment (mateuszkasprowicz) [reply sktime/base/_meta.py:724]: just to clarify what you suggest in practice is to introduce `__init__()` to `_HeterogenousMetaEstimator` and refactor class attributes `_steps_attr` and `_steps_fitted_attr` into tags set in the init method?

File: sktime/classification/compose/_pipeline.py (hunk 2 above)
Comment (mateuszkasprowicz) [reply sktime/classification/compose/_pipeline.py:171]: for vizualization - no, but I wanted to follow the convention described here: https://github.com/sktime/sktime/blob/3e1c92bbdceb0332ae6c12026528f650fa0ab4ec/sktime/base/_meta.py#L27-L31

File: sktime/base/_meta.py (hunk 1 above)
Comment (fkiraly) [reply sktime/base/_meta.py:724]: that is also an idea, but not what I mean - I meant, the value `"serial"` vs `"parallel"` passed to `_VisualBlock`, move that to be an attr or tag.

**Comment by mateuszkasprowicz** [comment]
the tags approach implemented, good idea!

a question about `TransformSelectForecaster` and `GroupbyCategoryForecaster`. in their not fitted state, they have a transfomer and a set of candidate forecasters. if an estimator is fitted we should display a transformer and a chosen forecaster as `serial`. but if it's not fitted I'd expect to see a transformer and a set of candidates parallelly. What do you think? @fkiraly 

note that to display them as `parallell` we would probably need another container object with a proper visual block kind.

**Comment by fkiraly** [comment]
> What do you think? @fkiraly

Hm, I thought in all cases the "blueprint" boxes are shown by the html repr, that is, the `__init__` arg. I do not think whether fitted or not should have an impact.

The two forecasters are pretty interesting. I would expect to see *all components* somehow displayed. I also see how the "natural" displays are not possible currently with the two variants.

Mockup of the "natural" displays:

`GroupbyCategoryForecaster`:
image

`TransformSelectForecaster` would be similar.

For ease of a first PR, I would just apply parallel and pool all estimators in it somehow.

**Comment by mateuszkasprowicz** [comment]
Can't find a way to display all forecasters next each other while maintaining serial display with transformer without creating a new class for the forecasters pool. I doubt it makes sense to introduce a new class just for the sake of visualization, so to me there are 2 options:
1. Display forecasters pool with a simple list and keep a serial display between transformer and forecasters
image
2. Display all estimators parallelly including the transformer
image

For me 1. makes much more sense for now. What do you think @fkiraly ?

**Comment by fkiraly** [comment]
Hm, interesting, @mateuszkasprowicz.

I would actually go with option 2 - it looks more professional, it clearly displays the components, and further allows users to inspect the components.

**Comment by mateuszkasprowicz** [comment]
both options are comparable. agreed that option 2 is more elegant, it lacks the proper flow though. But I guess that's something we have to live with.

with this, I'd say PR is complete. there's still a question of tests. [we discussed](https://github.com/sktime/sktime/issues/7222#issuecomment-2395448652) implementing an extra check for all _HeterogenousMetaEstimator descendants. Any guidance on this? @fkiraly

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks great!

Two points:
- I can see that a lot of tests are failing, but I cannot access the logs (internet security?) - could you summarize whether the failures are related?
- re the extra check, I think it might make sense, but the exact nature needs to be discussed. I would hence suggest to do that in a separate PR; if at all.

**Comment by mateuszkasprowicz** [comment]
@fkiraly fixed 'Permute' - also inherits from '_HeterogeneousMetaEstimator'; missed it initially

there's also a test for set_params failing for 'ColumnEnsembleTransformer' that I can't reproduce locally- can you take a look?

**Comment by fkiraly** [comment]
Will have a look.

**Review by fkiraly — CHANGES_REQUESTED** [review]
So far, I confirm this is a failure specifically introduced through this PR.

The likely reason is a modification or override to the `set_params` or `get_params` logic, from `_HeterogenousMetaEstimator`. Do you have a good idea what could cause this?

File: sktime/transformations/compose/_column.py [hunk 3]
```diff
@@ -127,13 +127,14 @@ class ColumnEnsembleTransformer(
...
         "handles-missing-data": True,
+        "visual_block_kind": "parallel",
     }
 
     # for default get_params/set_params from _HeterogenousMetaEstimator
     # _steps_attr points to the attribute of self
     # which contains the heterogeneous set of estimators
     # this must be an iterable of (name: str, estimator, ...) tuples for the default
-    _steps_attr = "transformers"
+    _steps_attr = "_transformers"
```
Comment (fkiraly) [inline sktime/transformations/compose/_column.py:137]: probably this is the reason?

Though, why does this not break with other trafos?

Comment (fkiraly) [reply sktime/transformations/compose/_column.py:137]: Does the `_transformers` attribute not delegate anywhere? I think that is the problem.

Comment (mateuszkasprowicz) [reply sktime/transformations/compose/_column.py:137]: only some checks fail on this which is peculiar - did you manage to reproduce this locally?

Comment (mateuszkasprowicz) [reply sktime/transformations/compose/_column.py]: couldn't figure it out neither with `_transformers` nor with `_transformers_plus_remainder`. to make it work with `set_params` it would likely need a tricky setter for this property or a bigger rework of the internals. For the sake of finally closing this PR I'd go with just overriding `_sk_visual_block` by adding `remainder` there. Perhaps, we can reference this here if we were to do a bigger refactor: 
https://github.com/sktime/sktime/issues/2537

On a more general note, I think it will be difficult to avoid overriding `_sk_visual_block` if we want to add some estimator details in visualizations (e.g., displaying column names selected for a given transformer after expanding the element in case of `ColumnEnsembleTransformer`.

Comment (fkiraly) [reply sktime/transformations/compose/_column.py]: ok, that makes sense. It would be appreciated if you open a new issue with the points that are not conclusively closed, so we can revisit them later.

Kudos, great PR!

Comment (mateuszkasprowicz) [reply sktime/transformations/compose/_column.py]: @fkiraly done, let me know if there's anything else needed before merging

**Review by fkiraly — APPROVED** [review]
Looks great! Sorry for the late review.
